oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/web/reactive/function/client/ServerOAuth2AuthorizedClientExchangeFilterFunction.java

```diff
@@ -251,27 +251,20 @@ public void setAccessTokenExpiresSkew(Duration accessTokenExpiresSkew) {
 
 	private Mono<OAuth2AuthorizedClient> refreshIfNecessary(ClientRequest request, ExchangeFunction next, OAuth2AuthorizedClient authorizedClient) {
 		ClientRegistration clientRegistration = authorizedClient.getClientRegistration();
-		if (isClientCredentialsGrantType(clientRegistration) && hasTokenExpired(authorizedClient)) {
```

Comment:
The logic for a `client_credentials` client is different compared to a client authorized via `authorization_code`. A `client_credentials` client will never have a refresh token but an `authorization_code` client may have a refresh token. Please review the [spec](https://tools.ietf.org/html/rfc6749#section-4.4.3):

> A refresh token SHOULD NOT be included